chain/network/src/peer/stream.rs

```diff
@@ -0,0 +1,203 @@
+#[rtype(result = "()")]
+pub(crate) enum Error {
+    #[error("send")]
+    Send(#[source] SendError),
+    #[error("recv")]
+    Recv(#[source] RecvError),
+}
+
+pub(crate) struct FramedStream {
+    queue_send: tokio::sync::mpsc::UnboundedSender<Vec<u8>>,
+    stats: Arc<connection::Stats>,
+    send_buf_size_metric: Arc<metrics::IntGaugeGuard>,
+    addr: actix::Addr<PeerActor>,
+}
+
+impl FramedStream {
+    pub fn spawn(
+        ctx: &mut actix::Context<PeerActor>,
+        peer_addr: SocketAddr,
+        stream: tokio::net::TcpStream,
+        stats: Arc<connection::Stats>,
+    ) -> Self {
+        let (tcp_recv, tcp_send) = tokio::io::split(stream);
+        let (queue_send, queue_recv) = tokio::sync::mpsc::unbounded_channel();
+        let send_buf_size_metric = Arc::new(metrics::MetricGuard::new(
+            &*metrics::PEER_DATA_WRITE_BUFFER_SIZE,
+            vec![peer_addr.to_string()],
+        ));
+        ctx.spawn(wrap_future({
+            let addr = ctx.address();
+            let stats = stats.clone();
+            let m = send_buf_size_metric.clone();
+            async move {
+                if let Err(err) = run_send_loop(tcp_send, queue_recv, stats, m).await {
+                    addr.do_send(Error::Send(SendError::IO(err)));
+                }
+            }
+        }));
+        ctx.spawn(wrap_future({
+            let addr = ctx.address();
+            let stats = stats.clone();
+            async move {
+                if let Err(err) = run_recv_loop(peer_addr, tcp_recv, addr.clone(), stats).await {
+                    addr.do_send(Error::Recv(err));
+                }
+            }
+        }));
+        Self { queue_send, stats, send_buf_size_metric, addr: ctx.address() }
+    }
+
+    /// Pushes `msg` to the send queue.
+    /// Silently drops message if it is too large,
+    /// or the connection has been closed.
+    /// Emits a critical error to PeerActor if send queue is full.
+    // TODO(gprusak): sending a too large message should probably be treated as a bug,
+    // since dropping messages may lead to hard-to-debug high-level issues.
+    // On the other hand silently dropping messages because of a closed queue should
+    // be OK, as it is indistinguishable from queueing messages on a closed connection.
+    pub fn send(&self, msg: Vec<u8>) {
+        if msg.len() > NETWORK_MESSAGE_MAX_SIZE_BYTES {
+            metrics::MessageDropped::InputTooLong.inc_unknown_msg();
+            return;
+        }
+        let mut buf_size =
+            self.stats.bytes_to_send.fetch_add(msg.len() as u64, Ordering::Acquire) as usize;
+        buf_size += msg.len();
+        self.stats.messages_to_send.fetch_add(1, Ordering::Acquire);
+        self.send_buf_size_metric.add(msg.len() as i64);
+        if buf_size > MAX_WRITE_BUFFER_CAPACITY_BYTES {
+            metrics::MessageDropped::MaxCapacityExceeded.inc_unknown_msg();
+            self.addr.do_send(Error::Send(SendError::QueueOverflow {
```

Comment:
Ah, so this is subtle -- we overflowed the queue here, so we want to kill the actor. But we also want to kill the write loop -- we don't want to spend the effort sending the messages already in the queue. We don't have any explicit early cancelation mechanism for the write loop, it get's canceld via drop of the future itself, b/c it is spawned on the actor's event loop. Probably worth a comment on the `Error`?

Reply:
ack, I've made it less subtle by queueing the message even if capacity is exceeded (stream::Error is still reported to PeerActor though). And I've made the docstring on stream::Error more elaborate.